lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp

```diff
@@ -238,5 +252,289 @@ bool circt::firrtl::fromJSON(json::Value &value,
+static void parseAugmentedType(
+    MLIRContext *context, DictionaryAttr augmentedType,
+    llvm::StringMap<llvm::SmallVector<Attribute>> &newAnnotations,
+    const StringRef &companion, StringAttr name, StringAttr defName) {
+
+  /// Unpack a ReferenceTarget encoded as a DictionaryAttr.  Return a pair
+  /// containing the Target string (up to the reference) and an array of
+  /// components.  The input DicionaryAttr encoding is a JSON object of a
+  /// serialized ReferenceTarget Scala class.  By example, this is
+  /// converting:
+  ///   ~Foo|Foo>a.b[0]
+  /// To:
+  ///   {"~Foo|Foo>a", {".b", "[0]"}}
+  /// The format of a ReferenceTarget object like:
+  ///   circuit: String
+  ///   module: String
+  ///   path: Seq[(Instance, OfModule)]
+  ///   ref: String
+  ///   component: Seq[TargetToken]
+  auto refTargetToString =
+      [&context](
+          DictionaryAttr refTarget) -> std::pair<std::string, ArrayAttr> {
+    auto circuit = refTarget.get("circuit").cast<StringAttr>().getValue();
+    auto module = refTarget.get("module").cast<StringAttr>().getValue();
+    auto path = refTarget.get("path").cast<ArrayAttr>();
+    if (!path.empty())
+      llvm_unreachable("Unable to handle non-local annotations in GrandCentral "
+                       "AugmentedTypes");
+    auto ref = refTarget.get("ref").cast<StringAttr>().getValue();
+    SmallVector<Attribute> componentAttrs;
+    for (auto component : refTarget.get("component").cast<ArrayAttr>()) {
+      auto dict = component.cast<DictionaryAttr>();
+      auto clazz = dict.get("class").cast<StringAttr>().getValue();
+      auto value = dict.get("value");
+      // A subfield like "bar" in "~Foo|Foo>foo.bar".
+      if (auto field = value.dyn_cast<StringAttr>()) {
+        assert(clazz == "firrtl.annotations.TargetToken$Field" &&
+               "A StringAttr target token must be found with a subfield target "
+               "token.");
+        componentAttrs.push_back(
+            StringAttr::get(context, (Twine(".") + field.getValue()).str()));
+        continue;
+      }
+      // A subindex like "42" in "~Foo|Foo>foo[42]".
+      if (auto index = value.dyn_cast<IntegerAttr>()) {
+        assert(clazz == "firrtl.annotations.TargetToken$Index" &&
+               "An IntegerAttr target token must be found with a subindex "
+               "target token.");
+        componentAttrs.push_back(StringAttr::get(
+            context,
+            (Twine("[") + index.getValue().toString(10, false) + "]").str()));
+        continue;
+      }
+      // Any other token shouldn't exist.
+      llvm_unreachable(
+          "Unexpected TargetToken. Only Field or Index should be possible.");
+    }
+    return {(Twine("~" + circuit + "|" + module + ">" + ref)).str(),
+            ArrayAttr::get(context, componentAttrs)};
+  };
+
+  static const char *package = "sifive.enterprise.grandcentral";
+
+  auto clazz = augmentedType.get("class").cast<StringAttr>();
```

Comment:
classAttr instead of clazz?

Reply:
Good thought. Will fix in a later rebase.